File: apps/pipeline/connectors/unstop.ts
// apps/pipeline/connectors/unstop.ts
// Unstop Puppeteer scraper — India-focused
// URL: https://unstop.com/hackathons
// CRITICAL: Filter to hackathons/build-a-thons ONLY — skip case studies, quizzes, debates
// 3s random delay between pages — high block risk (R1)
// Return PARTIAL on any block — never crash the pipeline

import puppeteer, { Browser } from 'puppeteer'
import type { IConnector, ConnectorResult, RawHackathon } from './base'

const KEYWORD_BLOCKLIST = [
  'quiz', 'case study', 'case-study', 'debate', 'trivia', 'essay',
  'management competition', 'moot court', 'business plan', 'quiz bowl',
]

const USER_AGENTS = [
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
  'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36',
  'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:124.0) Gecko/20100101 Firefox/124.0',
]

function randomUA(): string {
  return USER_AGENTS[Math.floor(Math.random() * USER_AGENTS.length)]
}

function randomDelay(): Promise<void> {
  // 3–4s random delay per page load
  return new Promise(r => setTimeout(r, 3000 + Math.random() * 1000))
}

function isBlocklisted(title: string, description: string): boolean {
  const text = `${title} ${description}`.toLowerCase()
  return KEYWORD_BLOCKLIST.some(kw => text.includes(kw))
}

interface ScrapedOpportunity {
  id?: string
  title: string
  organizer: string
  url: string
  deadline?: string
  startDate?: string
  prize?: string
  city?: string
  isOnline?: boolean
  tags: string[]
  description?: string
  teamMin?: number
  teamMax?: number
  logo?: string
}

export class UnstopConnector implements IConnector {
  source = 'UNSTOP'

  async fetch(): Promise<ConnectorResult> {
    const records: RawHackathon[] = []
    const errors: string[] = []
    let browser: Browser | null = null

    try {
      browser = await puppeteer.launch({
        headless: true,
        args: [
          '--no-sandbox',
          '--disable-setuid-sandbox',
          '--disable-dev-shm-usage',
          '--disable-gpu',
          '--no-first-run',
          '--disable-blink-features=AutomationControlled',
        ],
      })

      const page = await browser.newPage()
      await page.setUserAgent(randomUA())
      await page.setViewport({ width: 1280, height: 800 })

      // Mask automation signals
      await page.evaluateOnNewDocument(() => {
        Object.defineProperty(navigator, 'webdriver', { get: () => undefined })
      })

      // Block heavy resources
      await page.setRequestInterception(true)
      page.on('request', req => {
        if (['image', 'font', 'media', 'stylesheet'].includes(req.resourceType())) {
          req.abort()
        } else {
          req.continue()
        }
      })

      // Navigate to hackathons filter
      const baseUrl = 'https://unstop.com/hackathons?oppType=hackathon'

      let pageNum = 1
      const maxPages = 8

      while (pageNum <= maxPages) {
        const url = pageNum === 1 ? baseUrl : `${baseUrl}&page=${pageNum}`

        try {
          await page.goto(url, { waitUntil: 'networkidle2', timeout: 40000 })

          // Wait for opportunity cards
          const cardSelector = '.opportunity-card, un-opportunities-card, .single-card, [class*="card-wrapper"]'
          await page.waitForSelector(cardSelector, { timeout: 15000 }).catch(() => null)

          // Extra wait for dynamic content
          await new Promise(r => setTimeout(r, 2000))

          const scraped = await page.evaluate(() => {
            // Try multiple selectors since Unstop changes DOM frequently
            const cards = document.querySelectorAll(
              '.opportunity-card, .single-card, [data-ng-repeat*="opportunity"], .ng-scope.card'
            )

            const results: ScrapedOpportunity[] = []

            cards.forEach((card: Element) => {
              const titleEl = card.querySelector('h3, h2, .title, [class*="title"]')
              const orgEl = card.querySelector('.company-name, .org-name, [class*="company"], [class*="organizer"]')
              const linkEl = card.querySelector('a[href*="/p/"]') as HTMLAnchorElement | null
              const deadlineEl = card.querySelector('time, .deadline, [class*="date"], [class*="deadline"]')
              const prizeEl = card.querySelector('.prize, [class*="prize"], .winning-amount')
              const tagsEls = card.querySelectorAll('.tag, [class*="tag"], .category')
              const logoEl = card.querySelector('img') as HTMLImageElement | null
              const cityEl = card.querySelector('.city, [class*="location"], [class*="city"]')

              if (!titleEl) return

              const url = linkEl?.href ?? ''
              if (!url || !url.includes('unstop.com')) return

              results.push({
                title: titleEl.textContent?.trim() ?? '',
                organizer: orgEl?.textContent?.trim() ?? '',
                url,
                deadline: deadlineEl?.getAttribute('datetime') ?? deadlineEl?.textContent?.trim(),
                prize: prizeEl?.textContent?.trim(),
                city: cityEl?.textContent?.trim(),
                tags: Array.from(tagsEls).map((el: Element) => el.textContent?.trim() ?? '').filter(Boolean),
                logo: logoEl?.src,
                isOnline: !cityEl?.textContent?.trim(),
              })
            })
            return results
          }) as ScrapedOpportunity[]

          if (scraped.length === 0) {
            errors.push(`Unstop page ${pageNum}: no cards found (possible block or page structure change)`)
            break
          }

          for (const h of scraped) {
            // Skip blocklisted events
            if (isBlocklisted(h.title, h.description ?? '')) {
              continue
            }

            const record = this.normalize(h)
            if (record) records.push(record)
          }

          pageNum++
          await randomDelay()

        } catch (pageErr) {
          errors.push(`Unstop page ${pageNum} error: ${pageErr}`)
          // Don't crash — return what we have as PARTIAL
          break
        }
      }

    } catch (err) {
      errors.push(`Unstop browser launch error: ${err}`)
      if (records.length === 0) {
        return { source: this.source, records: [], errors, status: 'FAILED' }
      }
    } finally {
      if (browser) {
        await browser.close().catch(() => {})
      }
    }

    return {
      source: this.source,
      records,
      errors,
      status: records.length === 0 ? 'FAILED' : errors.length === 0 ? 'SUCCESS' : 'PARTIAL',
    }
  }

  private normalize(h: ScrapedOpportunity): RawHackathon | null {
    if (!h.title || !h.url) return null

    // Extract sourceId from URL: /p/some-hackathon-123 → 123 or slug
    const idMatch = h.url.match(/\/p\/([^/?#]+)-(\d+)$/) ?? h.url.match(/\/p\/([^/?#]+)/)
    const sourceId = idMatch?.[2] ?? idMatch?.[1] ?? h.url.split('/p/')[1] ?? h.url

    // Parse deadline
    let registrationClose: string
    try {
      if (h.deadline) {
        // Handle formats like "Apr 30, 2025" or ISO
        registrationClose = new Date(h.deadline).toISOString()
        if (isNaN(new Date(registrationClose).getTime())) return null
      } else {
        return null // No deadline = quality gate will reject anyway
      }
    } catch {
      return null
    }

    // Parse prize
    let prizePool: number | undefined
    if (h.prize) {
      const cleaned = h.prize.replace(/[^0-9.]/g, '')
      const parsed = parseFloat(cleaned)
      if (!isNaN(parsed) && parsed > 0) prizePool = parsed
    }

    let mode: 'ONLINE' | 'OFFLINE' | 'HYBRID' = 'ONLINE'
    if (h.city && !h.isOnline) mode = 'OFFLINE'

    return {
      sourceId,
      title: h.title,
      organizerName: h.organizer || 'Unstop Hackathon',
      organizerLogoUrl: h.logo,
      applyUrl: h.url,
      registrationClose,
      mode,
      prizePool,
      prizeCurrency: 'INR',
      prizeDescription: h.prize,
      themeTags: h.tags.slice(0, 8),
      scope: 'INDIA',
      indiaRegion: h.city ?? 'Pan-India',
      eligibility: 'OPEN',
      durationType: 'CUSTOM',
      teamSizeMin: h.teamMin ?? 1,
      teamSizeMax: h.teamMax,
      sponsors: [],
    }
  }
}

File: pipeline/connectors/unstop.py
"""
unstop.py — Unstop connector. JS-rendered Angular SPA.
Method: ZeroCrawl (browser mode with anti-detection).
India-focused. Filters to hackathons only.
"""
import re
from bs4 import BeautifulSoup

from .base import BaseConnector, ConnectorResult, RawHackathon

LIST_URL = "https://unstop.com/hackathons"
BLOCKLIST = {"quiz", "case study", "debate", "trivia", "essay", "moot court", "management", "contest"}


def _is_blocked(title: str) -> bool:
    tl = title.lower()
    return any(b in tl for b in BLOCKLIST)


def _parse_prize(text: str):
    if not text:
        return None
    text = text.replace(",", "").replace("₹", "").replace("$", "").replace("USD", "")
    nums = re.findall(r"\d+(?:\.\d+)?", text)
    if nums:
        try:
            v = float(nums[0])
            if "l" in text.lower() and v < 1000:
                v *= 100_000
            return v
        except ValueError:
            pass
    return None


def _parse_html(html: str, records: list) -> int:
    """Parse hackathon cards from Unstop HTML. Returns count of cards found."""
    soup = BeautifulSoup(html, "html.parser")
    count = 0

    # Try various card selectors
    cards = (
        soup.select("div[class*='opportunity']") or
        soup.select("div[class*='card-wrapper']") or
        soup.select("app-opportunity-card") or
        soup.select("div[class*='listing__card']") or
        soup.select("div[class*='single-opportunity']")
    )

    if cards:
        for card in cards:
            try:
                title_el = card.select_one("h3, h2, [class*='title'], [class*='name'], [class*='opportunity-name']")
                if not title_el:
                    continue
                title = title_el.get_text(strip=True)
                if not title or _is_blocked(title):
                    continue

                link_el = card.select_one("a[href]")
                if not link_el:
                    continue
                href = link_el.get("href", "")
                apply_url = href if href.startswith("http") else f"https://unstop.com{href}"
                slug = href.rstrip("/").split("/")[-1]

                prize_el = card.select_one("[class*='prize'], [class*='reward'], [class*='amount'], [class*='winning'], [class*='cash']")
                prize_text = prize_el.get_text(strip=True) if prize_el else ""
                prize = _parse_prize(prize_text)

                org_el = card.select_one("[class*='org'], [class*='company'], [class*='host'], [class*='organizer'], [class*='institute']")
                org = org_el.get_text(strip=True) if org_el else "Unstop"

                tags_els = card.select("[class*='tag'], [class*='category'], [class*='track']")
                tags = [t.get_text(strip=True) for t in tags_els[:5] if t.get_text(strip=True)]

                records.append(RawHackathon(
                    source_id=f"unstop-{slug}",
                    title=title,
                    organizer_name=org,
                    apply_url=apply_url,
                    registration_close=None,
                    prize_pool=prize,
                    prize_currency="INR",
                    theme_tags=tags,
                    mode="ONLINE",
                    scope="INDIA",
                ))
                count += 1
            except Exception:
                continue
    else:
        # Fallback: grab all hackathon links
        links = soup.select("a[href*='/hackathon'], a[href*='/o/']")
        seen: set = set()
        for link in links:
            try:
                href = link.get("href", "")
                if not href or href in seen:
                    continue
                if href in ["/hackathons", "https://unstop.com/hackathons"]:
                    continue
                seen.add(href)
                apply_url = href if href.startswith("http") else f"https://unstop.com{href}"
                raw_title = link.get("aria-label") or link.get_text(strip=True)
                title = raw_title.split('\n')[0].strip() if raw_title else ""
                if not title or len(title) < 4 or _is_blocked(title):
                    continue
                slug = href.rstrip("/").split("/")[-1]
                records.append(RawHackathon(
                    source_id=f"unstop-{slug}",
                    title=title,
                    organizer_name="Unstop",
                    apply_url=apply_url,
                    registration_close=None,
                    mode="ONLINE",
                    scope="INDIA",
                ))
                count += 1
            except Exception:
                continue

    return count


class UnstopConnector(BaseConnector):
    SOURCE = "UNSTOP"
    SCOPE = "INDIA"

    def fetch(self) -> ConnectorResult:
        from ..zerocrawl_bridge import fetch_js_page

        records = []
        error = None

        try:
            print(f"[{self.SOURCE}] Fetching {LIST_URL} via ZeroCrawl browser mode...")
            # Unstop is an Angular SPA — force browser mode
            html = fetch_js_page(
                LIST_URL,
                wait_for_selector="div[class*='opportunity'], app-opportunity-card",
                timeout=90,
            )

            if not html:
                error = "ZeroCrawl returned empty response"
                print(f"[{self.SOURCE}] {error}")
            else:
                count = _parse_html(html, records)
                print(f"[{self.SOURCE}] Parsed {count} hackathons")

        except Exception as e:
            error = str(e)
            print(f"[{self.SOURCE}] Error: {e}")

        print(f"[{self.SOURCE}] Total: {len(records)} records")
        status = "SUCCESS" if records else ("PARTIAL" if error else "FAILED")
        return ConnectorResult(source=self.SOURCE, records=records, status=status, error=error)
